src/NuGet.Core/NuGet.DependencyResolver.Core/GraphModel/GraphOperations.cs

```diff
@@ -176,179 +206,460 @@ private static bool TryResolveConflicts<TItem>(this GraphNode<TItem> root, List<
+        {
+            var count = innerNodes.Count;
+            for (var i = 0; i < count; i++)
+            {
+                var innerNode = innerNodes[i];
+                queue.Enqueue(NodeWithState.Create(innerNode, innerState));
+            }
+        }
+
+        private static void AddInnerNodesToQueue<TItem>(IList<GraphNode<TItem>> innerNodes, Queue<GraphNode<TItem>> queue)
+        {
+            var count = innerNodes.Count;
+            for (var i = 0; i < count; i++)
+            {
+                var innerNode = innerNodes[i];
+                queue.Enqueue(innerNode);
+            }
+        }
+
+        [ThreadStatic]
+        private static Dictionary<GraphNode<RemoteResolveResult>, GraphNode<RemoteResolveResult>> _tempDowngrades;
+
+        public static Dictionary<GraphNode<RemoteResolveResult>, GraphNode<RemoteResolveResult>> RentDowngradesDictionary()
+        {
+            var dictionary = _tempDowngrades;
+            if (dictionary != null)
+            {
+                _tempDowngrades = null;
+                return dictionary;
+            }
+
+            return new Dictionary<GraphNode<RemoteResolveResult>, GraphNode<RemoteResolveResult>>();
+        }
+
+        public static void ReleaseDowngradesDictionary(Dictionary<GraphNode<RemoteResolveResult>, GraphNode<RemoteResolveResult>> dictionary)
+        {
+            if (_tempDowngrades == null)
+            {
+                dictionary.Clear();
+                _tempDowngrades = dictionary;
+            }
+        }
+
+        private static class Cache<TItem, TState>
+        {
+            [ThreadStatic]
+            private static Queue<NodeWithState<TItem, TState>> _queue;
+
+
+            public static Queue<NodeWithState<TItem, TState>> RentQueue()
+            {
+                var queue = _queue;
+                if (queue != null)
                 {
-                    visitor(node);
-                    return 0;
-                });
+                    _queue = null;
+                    return queue;
+                }
+
+                return new Queue<NodeWithState<TItem, TState>>();
+            }
+
+            public static void ReleaseQueue(Queue<NodeWithState<TItem, TState>> queue)
+            {
+                if (_queue == null)
+                {
+                    queue.Clear();
+                    _queue = queue;
+                }
+            }
+
+        }
+
+        private static class Cache<TItem>
+        {
+            [ThreadStatic]
+            private static Queue<GraphNode<TItem>> _queue;
+            [ThreadStatic]
+            private static Dictionary<string, GraphNode<TItem>> _dictionary;
+            [ThreadStatic]
+            private static Tracker<TItem> _tracker;
+
+            public static Queue<GraphNode<TItem>> RentQueue()
+            {
+                var queue = _queue;
+                if (queue != null)
+                {
+                    _queue = null;
+                    return queue;
```

Comment:
Is this guaranteed to be thread safe?  There will be a lot of different projects restoring on different threads using this same static.  I'm not very familiar with ThreadStatic to know how this is handled.